src/NodaTime/DateInterval.cs

```diff
@@ -204,5 +246,108 @@ public override string ToString()
+            /// 10000-12-31), which cannot be represented as a DateInterval.  Instead, this method maps all non-empty
+            /// intervals to inclusive intervals, and retains all empty intervals as empty (and exclusive by definition)
+            /// intervals.
+            /// </para>
+            /// </remarks>
+            /// <returns>The normalized interval.</returns>
+            [NotNull]
+            internal static DateInterval Normalize([NotNull] DateInterval obj)
+            {
+                return obj.Inclusive || obj.Length == 0 ? obj : new DateInterval(obj.Start, obj.End.PlusDays(-1), true);
+            }
+        }
+
+        /// <summary>
+        /// Equality comparer that considers the dates containing within intervals.
+        /// </summary>
+        private sealed class ContainedDatesDateIntervalEqualityComparer : EqualityComparer<DateInterval>
+        {
+            internal static readonly ContainedDatesDateIntervalEqualityComparer Instance = new ContainedDatesDateIntervalEqualityComparer();
+            private static readonly DateInterval CanonicalIsoEmptyInterval = CreateCanonicalEmptyInterval(CalendarSystem.Iso);
+
+            private ContainedDatesDateIntervalEqualityComparer()
+            {
+            }
+
+            public override bool Equals(DateInterval x, DateInterval y)
+            {
+                if (ReferenceEquals(x, y))
+                {
+                    return true;
+                }
+                if (ReferenceEquals(x, null) || ReferenceEquals(y, null))
+                {
+                    return false;
+                }
+                return Normalize(x) == Normalize(y);
+            }
+
+            public override int GetHashCode([NotNull] DateInterval obj) =>
+                Normalize(Preconditions.CheckNotNull(obj, nameof(obj))).GetHashCode();
+
+            /// <summary>
+            /// Returns a normalized version of this interval, by converting non-empty exclusive intervals to inclusive
+            /// ones, and canonicalizing empty intervals to the same start date.
+            /// </summary>
+            /// <returns>The normalized interval.</returns>
+            [NotNull]
+            private static DateInterval Normalize([NotNull] DateInterval obj)
+            {
+                if (obj.Length > 0)
+                {
+                    return NormalizingDateIntervalEqualityComparer.Normalize(obj);
+                }
+                var calendar = obj.Start.Calendar;
+                return calendar == CalendarSystem.Iso ? CanonicalIsoEmptyInterval : CreateCanonicalEmptyInterval(calendar);
+            }
+
+            private static DateInterval CreateCanonicalEmptyInterval(CalendarSystem calendar)
+            {
+                // This can use any arbitrary date for the given calendar, so long as it's a valid date.
+                var date = new LocalDate(calendar.MinYear, 1, 1, calendar);
```

Comment:
I had to mentally check that this really *would* be valid, even in calendars where the first month of the year isn't 1, but I think it is...